provider/core/src/refcount.rs

```diff
@@ -0,0 +1,144 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::ffi::c_void;
+use core::fmt::Debug;
+use core::mem::size_of;
+use core::{mem::ManuallyDrop, ptr::NonNull};
+use alloc::sync::Arc;
+use alloc::{string::String, rc::Rc};
+use alloc::boxed::Box;
+use stable_deref_trait::{StableDeref, CloneStableDeref};
+
+type WrappedType = Option<Box<[u8]>>;
+
+// Safe as a sentinel because 0x1 is not a valid usize address
+const SENTINEL: *const WrappedType = 1usize as *const WrappedType;
+
+/// A thin wrapper over [`NonNull`] that works with `*const T`
+#[derive(Debug)]
+#[repr(transparent)]
+pub(crate) struct NonNullConst<T>(NonNull<T>);
+
+impl<T> NonNullConst<T> {
+    #[inline]
+    pub const fn new_sentinel() -> Self {
+        // Safety: `SENTINEL` is non-null
+        Self(unsafe { NonNull::new_unchecked(SENTINEL as *mut T) })
+    }
+    #[inline]
+    pub fn new_from_rc(rc: Rc<T>) -> Self {
+        // Safety: `Rc::into_raw` returns a non-null pointer
+        Self(unsafe { NonNull::new_unchecked(Rc::into_raw(rc) as *mut T) })
```

Comment:
nit: if this type just exists to encapsulate non-null `*const`, then the Rc logic should not be here. There should be a constructor from `*const`, which can be called with `SENTINEL`, `Rc::into_raw(rc)`, `Arc::into_raw(rc)`, etc.

Reply:
It's all merged down into a single type now.